lib/headers.c

```diff
@@ -335,6 +335,7 @@ CURLcode Curl_headers_push(struct Curl_easy *data, const char *header,
 static void headers_init(struct Curl_easy *data)
 {
   Curl_llist_init(&data->state.httphdrs, NULL);
+  data->state.requests = 0;
```

Comment:
is there some reason this can't be put in pretransfer?

Reply:
No. I couldn't quite make up my mind where it belongs. But I take it you think `Curl_pretransfer` is a better place? Curl_headers_cleanup() is called from pretransfer and this headers_init() is called from Curl_headers_cleanup().

Reply:
Yes but I see what you are saying. That's sort of hard to figure out. I suggest don't call init implicitly from cleanup, I would actually do it reverse:

~~~diff
 /*
  * Curl_headers_init(). Init the headers subsystem.
  */
-static void headers_init(struct Curl_easy *data)
+CURLcode Curl_headers_init(struct Curl_easy *data)
 {
+  CURLcode rc = Curl_headers_cleanup(data);
+  if(rc)
+    return rc;
   Curl_llist_init(&data->state.httphdrs, NULL);
+  return CURLE_OK;
 }
~~~
 then in pretransfer

result = Curl_headers_init(
if (result)
error

data->state.requests = 0;
data->state.followlocation = 0; /* reset the location-follow counter */
data->state.this_is_a_follow = FALSE; /* reset this */

etc

Though requests is only used by headers functions right now it could cause problems not to have it in pretransfer if something code is added that evaluates it elsewhere

Reply:
Maybe, I just think that's a little much to "just" fix this particular bug. I rather do the smaller change now and move the init to Curl_pretransfer. See my follow-up commit for this.